source/s3_copy_object.c

```diff
@@ -0,0 +1,728 @@
+        }
+
+        /* If the number of parts completed is less than the number of parts sent, then we need to wait until all of
+         * those parts are done sending before aborting. */
+        if (copy_object->synced_data.num_parts_completed < copy_object->synced_data.num_parts_sent) {
+            goto has_work_remaining;
+        }
+
+        /* If the complete-multipart-upload is already in flight, then we can't necessarily send an abort. */
+        if (copy_object->synced_data.complete_multipart_upload_sent &&
+            !copy_object->synced_data.complete_multipart_upload_completed) {
+            goto has_work_remaining;
+        }
+
+        /* If the complete-multipart-upload completed successfully, then there is nothing to abort since the transfer
+         * has already finished. */
+        if (copy_object->synced_data.complete_multipart_upload_completed &&
+            copy_object->synced_data.complete_multipart_upload_error_code == AWS_ERROR_SUCCESS) {
+            goto no_work_remaining;
+        }
+
+        /* If we made it here, and the abort-multipart-upload message hasn't been sent yet, then do so now. */
+        if (!copy_object->synced_data.abort_multipart_upload_sent) {
+            if (copy_object->upload_id == NULL) {
+                goto no_work_remaining;
+            }
+
+            request = aws_s3_request_new(
+                meta_request,
+                AWS_S3_COPY_OBJECT_REQUEST_TAG_ABORT_MULTIPART_UPLOAD,
+                0,
+                AWS_S3_REQUEST_FLAG_RECORD_RESPONSE_HEADERS | AWS_S3_REQUEST_FLAG_ALWAYS_SEND);
+
+            copy_object->synced_data.abort_multipart_upload_sent = true;
+
+            goto has_work_remaining;
+        }
+
+        /* Wait for the multipart upload to be completed. */
+        if (!copy_object->synced_data.abort_multipart_upload_completed) {
+            goto has_work_remaining;
+        }
+
+        goto no_work_remaining;
+    }
+
+has_work_remaining:
+    work_remaining = true;
+
+no_work_remaining:
+
+    if (!work_remaining) {
+        aws_s3_meta_request_set_success_synced(meta_request, AWS_S3_RESPONSE_STATUS_SUCCESS);
+    }
+
+    aws_s3_meta_request_unlock_synced_data(meta_request);
+
+    if (work_remaining) {
+        *out_request = request;
+    } else {
+        AWS_ASSERT(request == NULL);
+
+        aws_s3_meta_request_finish(meta_request);
+    }
+
+    return work_remaining;
+}
+
+/* Given a request, prepare it for sending based on its description. */
+static int s_s3_copy_object_prepare_request(struct aws_s3_meta_request *meta_request, struct aws_s3_request *request) {
+    AWS_PRECONDITION(meta_request);
+
+    struct aws_s3_copy_object *copy_object = meta_request->impl;
+    AWS_PRECONDITION(copy_object);
+
+    aws_s3_meta_request_lock_synced_data(meta_request);
+
+    struct aws_http_message *message = NULL;
+
+    switch (request->request_tag) {
+
+        case AWS_S3_COPY_OBJECT_REQUEST_TAG_GET_OBJECT_SIZE: {
```

Comment:
reasoning for what each case means above each case statement would be helpful

Reply:
Added comments for each case label, thanks!